crates/common/tedge_config/src/tedge_config_cli/new_tedge_config.rs

```diff
@@ -0,0 +1,1412 @@
+///
+/// // Getters for settings with defaults are infallible
+/// assert_eq!(config.mqtt_port(), 1883);
+///
+/// // Getters for other settings will fail if the value is not setting
+/// assert!(config.mqtt_external_port().is_err());
+///
+/// // The defaults for some values are inferred from the config location
+/// assert!(config.device_key_path().starts_with(temp_dir.utf8_path()));
+///
+/// # Ok::<_, TEdgeConfigError>(())
+/// ```
+pub struct NewTEdgeConfig {
+    /// The configured values from tedge.toml or environment variables
+    stored: TEdgeConfigDto,
+    /// A cache for the device id, to save us re-reading certificates unneccessarily
+    device_id: OnceCell<String>,
+    /// The location of the configuration file
+    config_location: TEdgeConfigLocation,
+}
+
+impl NewTEdgeConfig {
+    pub(crate) fn new(value: crate::TEdgeConfig, location: &TEdgeConfigLocation) -> Self {
+        Self {
+            stored: value.data,
+            device_id: <_>::default(),
+            config_location: location.to_owned(),
+        }
+    }
+
+    /// Read an arbitrary key as a string
+    ///
+    /// This is generally useful for the `tedge` CLI. Other crates should use
+    /// the getter methods documented below, such as [device_id](Self::device_id),
+    /// [c8y_url](Self::c8y_url) and [az_mapper_timestamp](Self::az_mapper_timestamp).
+    pub fn read(&self, key: ReadableKey) -> Result<Option<String>, ConfigSettingError> {
+        use ReadOnlyKey::*;
+        use ReadableKey::*;
+
+        Ok(match key {
+            ReadOnly(DeviceId) => Some(self.device_id()?.to_string()),
+            ReadOnly(HttpAddress) => Some(self.http_address().to_string()),
+            ReadWrite(key) => self.read_configured_value_string(key),
+        })
+    }
+
+    /// Reads `device.id` from the configured certificate
+    ///
+    /// # Errors
+    /// This returns an error if the certificate path cannot be read
+    /// ([ConfigSettingError::ReadOnlySettingNotConfigured]).
+    ///
+    /// It also returns an error if the PEM file cannot be parsed
+    /// ([ConfigSettingError::DerivationFailed]).
+    pub fn device_id(&self) -> Result<&str, ConfigSettingError> {
+        self.device_id
+            .get_or_try_init(|| {
+                let cert_path = self.stored.device().cert_path(&self.config_location);
+                PemCertificate::from_pem_file(&*cert_path)?.subject_common_name()
+            })
+            .map(|s| s.as_str())
+            .map_err(|err| cert_error_into_config_error("device.id", err))
+    }
+
+    /// Getter for `http.address`
+    pub fn http_address(&self) -> IpAddress {
+        self.stored
+            .mqtt()
+            .external_bind_address()
+            .unwrap_or_else(|| self.stored.mqtt().bind_address())
+    }
+
+    /// Getter for `firmware.child_update_timeout`
+    pub fn firmware_child_update_timeout(&self) -> Duration {
+        self.stored.firmware().child_update_timeout()
+    }
```

Comment:
This is a special case because it felt sensible for any Rust code accessing this to access this as a `Duration`, but `tedge config` needs to read and write this as a number of seconds. So when we read the string value, we get the number of seconds, but when we read the typed value, we get a `Duration`.

Reply:
I would prefer to address the issue at the `TEdgeConfigDto` level using `#[serde(deserialize_with = "to_duration", serialize_with = "from_duration")]`.

Reply:
In which case we actually would need a wrapper type because it needs ToString and FromStr too for tedge config